### DISCUSSION TITLE
For which reasons can a codefix be unavailable in VS?

### DISCUSSION THREAD

**Discussion by taori** [dotnet/roslyn#64881] [Q&A]
I am currently working on some package to simplify the creation of CLI programs. To guide the library usage and simplify some processes i added a couple of features using sourcegenerators, analyzers and now also codefixes. Everything but the codefix work. 

My code is here:

https://github.com/taori/Amusoft.Toolkit.System.CommandLine/tree/codefix-bindhandler/src/Amusoft.Toolkit.System.CommandLine.Generator

What i have tried so far is:
- delete .vs folder (fixes a lot of generator issues)
- delete bin/obj
- insert into fresh analyzer with codefix scaffold (works just fine when installed as vsix)
- upgrade workspace in fresh project - works too
- check other projects for maybe different configuration to create nupkg files with analyzer+codefixes

Is there some tracing i can enable to figure out why my codefix won't load with the current package? 
Can i not ship Analyzers together with codefixes in the same nupkg?

I am unsure how to move further with this. Any help is welcome

**Comment by CyrusNajmabadi** [comment]
@jmarolf Are CodeFixes supported in nupkgs?  or just vsixes?  Thanks!

**Reply by taori** [reply]
More specifically i am wondering if perhaps combining both in one assembly is a technical issue right now - i remember CodeFix+Analyzer being part of one project was an older state of the scaffolding template, but with the current scaffolding template they are split in 2 projects. Is this a necessity now to be discovered properly?

**Reply by jmarolf** [reply]
code fixers have been supported in nuget packages since roslyn 1.0. Code _refactorings_ are not.

**Reply by jmarolf** [reply]
> More specifically i am wondering if perhaps combining both in one assembly is a technical issue right now -

Yes, your analyzer gets loaded by the compiler so the compiler needs to be able to load all your dependencies. a codefix and analyzer living in the same assembly means that the compiler will attempt to load the IDE dlls (since your assembly has an explicit dependency on them) and that will fail.

**Reply by taori** [reply]
![grafik](https://user-images.githubusercontent.com/5545184/200170255-d39a0687-e871-4a65-b2d2-b3f23a931bd5.png)
Shouldn't this be fine then? There is no reference to the analyzer assembly. They are both in their own project with their own references, just shipped together in one nupkg (similar to StyleCop.Analyzers) For me the biggest issue here is the lack of information, that something about assembly loading isn't working the way it should. Is there any logs i can enable in VS to get more details about what is wrong?

**Comment by 333fred** [comment]
You should not combine your analyzers/generators and your codefixers in a single assembly. Analyzers and generators cannot depend on the Workspaces layer, while codefixers _must_ depend on the Workspaces layer.

**Reply by taori** [reply]
Whatever the root cause is seems to go deeper than just fixing it by seperating analyzer+codefix. I have isolated the problem in a sample repo: https://github.com/taori/ReproRepo/tree/main/codeFixMissing 

repro through: 
## works
- open Analyzer1.sln, start vsix 
- open cli.standalone.sln. 
- TestCommand should provide codefix (Inject BindHandler call) - that part works.

## does not work:

- open codeFixMissing.sln
- see nupkg Amusoft.Toolkit.System.CommandLine.Generator.2.0.0-alpha.1.73.nupkg (based on [this](https://github.com/taori/Amusoft.Toolkit.System.CommandLine/commit/37c26229c57852bc645ef25cbd6b600213ae394f) commit)
- analyzer points out issue - but no codefix available

Any information on how to access logs to trace down the issue why the codefix won't show are welcome. I have looked at [this](https://www.nuget.org/packages/StyleCop.Analyzers/1.2.0-beta.435) for reference to see how a package that does ship analyzer+codefix is structured and tried to replicate it. (no nuspec reference, copy to analyzer subfolder)
